### PR TITLE
FSDP example

### PR DESCRIPTION
**Pull request by HamidShojanazeri**
This example shows training a HF T5 model with FSDP to be used with its [tutorial](https://github.com/pytorch/tutorials/pull/1959#discussion_r913870745)

### CODE REVIEW

File: FSDP/T5_training.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+bfSixteen = MixedPrecision(
+    param_dtype=torch.bfloat16,
+    # Gradient communication precision.
+    reduce_dtype=torch.bfloat16,
+    # Buffer precision.
+    buffer_dtype=torch.bfloat16,
+)
+
+fp32_policy = MixedPrecision(
```
Comment (msaroufim) [inline FSDP/T5_training.py:81]: seems ununsed? same comment for fpsixteen

File: FSDP/T5_training.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        transformer_auto_wrap_policy,
+        transformer_layer_cls={
+            T5Block,
+        },
+    )
+    sharding_strategy: ShardingStrategy = ShardingStrategy.SHARD_GRAD_OP #for Zero2 and FULL_SHARD for Zero3
+    torch.cuda.set_device(local_rank)
+   
+   
+    #init_start_event = torch.cuda.Event(enable_timing=True)
```
Comment (msaroufim) [inline FSDP/T5_training.py:209]: let's remove the comments

File: FSDP/T5_training.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    if bf16_ready:
+        mp_policy = bfSixteen
+    else:
+        mp_policy = None # defaults to fp32
+        
+    model = FSDP(model,
+        auto_wrap_policy=t5_auto_wrap_policy,
+        mixed_precision=mp_policy,
+        #sharding_strategy=sharding_strategy,
+        device_id=torch.cuda.current_device())
```
Comment (msaroufim) [inline FSDP/T5_training.py:231]: I see `--no-cuda` argument later down in the CLI. Should users be able to configure all of these options with CLI and if not let's remove the CLI options that they don't need to know about

File: FSDP/T5_training.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+                mem_reserved_tracker.append(
+                    format_metrics_to_gb(torch.cuda.memory_reserved())
+                )
+            print(f"completed save and stats zone...")
+        
+
+    #init_end_event.record()
+        
+    #if rank == 0:
+        #print(f"Cuda event elapsed time: {init_start_event.elapsed_time(init_end_event) / 1000}sec")
```
Comment (msaroufim) [inline FSDP/T5_training.py:281]: please remove the commented out code

File: FSDP/T5_training.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                cpu_state = model.state_dict()
+            #print(f"saving process: rank {rank}  done w state_dict")
+           
+
+            if rank == 0:
+                print(f"--> saving model ...")
+                currEpoch = (
+                    "-" + str(epoch) + "-" + str(round(curr_val_loss.item(), 4)) + ".pt"
+                )
+                print(f"--> attempting to save model prefix {currEpoch}")
```
Comment (msaroufim) [inline FSDP/T5_training.py:303]: nit: saving could be its own function

File: FSDP/T5_training.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+from summarization_dataset import *
+from transformers.models.t5.modeling_t5 import T5Block
+from typing import Type
+import time
+import tqdm
+from datetime import datetime
+
+g_gigabyte = 1024**3
+
+def setup():
```
Comment (msaroufim) [inline FSDP/T5_training.py:49]: nit: since these 2 functions are both 1 liners best to just inline them in code directly

File: FSDP/README.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+## FSDP T5
+
+To run the T5 example with FSDP for text summarization:
```
Comment (msaroufim) [inline FSDP/README.md:3]: Link to T5 and FSDP documentation

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: FSDP/T5_training.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+    and torch.cuda.is_bf16_supported()
+    and LooseVersion(torch.version.cuda) >= "11.0"
+    and dist.is_nccl_available()
+    and nccl.version() >= (2, 10)
+    )
+    
+    if bf16_ready:
+        mp_policy = bfSixteen
+    else:
+        mp_policy = None # defaults to fp32
```
Comment (hudeven) [inline FSDP/T5_training.py:225]: @HamidShojanazeri How about wrapping these lines and line 65-85 into a function? e.g. mp_policy = get_mp_policy(...)

**Review by hudeven — CHANGES_REQUESTED** [review]
@HamidShojanazeri I think a good FSDP example would be very useful for users doing large scale training. Thanks for your contribution! Requesting to resolve the comments.

**Comment by msaroufim** [comment]
@rohan-varma @lessw2020 @HamidShojanazeri once you tell @hudeven and I that you'd like to merge the PR let us know. This has been open for a while.

**Comment by lessw2020** [comment]
Let me review - I was not even aware this PR existed until today, so thanks for the direct link.

File: FSDP/T5_training.py (hunk 8 above)
Comment (lessw2020) [reply FSDP/T5_training.py:225]: I have done this in a different codebase and agree it should be modularized.

File: FSDP/T5_training.py (hunk 5 above)
Comment (lessw2020) [reply FSDP/T5_training.py:303]: agree - it's done this way in a different fsdp repo I have.

File: FSDP/README.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+~~~
+
+## Install the requirements:
+~~~
+pip install -r requirements.txt
+~~~
+## Ensure you are running a recent version of PyTorch:
+see https://pytorch.org to install at least 1.12 and ideally a current nightly build. 
+
+Start the training with Torchrun (adjust nproc_per_node to your GPU count):
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply FSDP/README.md:20]: @msaroufim - atm FSDP does not work at all on CPU as it's totally reliant on Cuda streams for comms and compute.

**Comment by lessw2020** [comment]
General comment - this example does not use activation checkpointing due to the timing of this PR (it wasn't added in FSDP until after this PR).  
But I think it would be good to update this example with it, to make sure it's present as activation checkpointing is one of our biggest throughput boosters.

File: FSDP/README.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+## FSDP T5
+
+To run the T5 example with FSDP for text summarization:
+
+## Get the wikihow dataset
```
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply FSDP/README.md]: done

File: FSDP/T5_training.py (hunk 3 above)
Comment (HamidShojanazeri) [reply FSDP/T5_training.py]: updated

File: FSDP/T5_training.py (hunk 5 above)
Comment (HamidShojanazeri) [reply FSDP/T5_training.py]: updated

**Comment by HamidShojanazeri** [comment]
@msaroufim , @hudeven sorry for the delay I addressed the comments and made the code more modular, would be great if we could merge this.

**Comment by HamidShojanazeri** [comment]
> General comment - this example does not use activation checkpointing due to the timing of this PR (it wasn't added in FSDP until after this PR). But I think it would be good to update this example with it, to make sure it's present as activation checkpointing is one of our biggest throughput boosters.

Added the AC and rate_lmiter as well+ model checkpointings.

**Review by msaroufim — APPROVED** [review]

**Comment by msaroufim** [comment]
@svekars any idea if the doc build is flaking for any reason?

@HamidShojanazeri do you mind rebasing on main to see if the error goes away

**Pull request merged**